### PR TITLE
Avoid clear locator cache during invocation retries

### PR DESCRIPTION
**Pull request by pepone**
This PR reworks the computation of indirect endpoints to allow requesting a fresh endpoint without clearing the cache

### CODE REVIEW

File: csharp/src/Ice/LocatorInfo.cs [hunk 1]
```diff
@@ -90,83 +90,81 @@ internal void ClearCache(Reference reference)
...
 
             Location location = reference.Location;
-
             if (reference.IsWellKnown)
             {
                 // First, we check the cache.
-                (endpoints, location, cached) = GetResolvedWellKnownProxyFromCache(reference,
-                                                                                   reference.LocatorCacheTimeout);
-                if (!cached)
+                (endpoints, location, endpointsAge) = GetResolvedWellKnownProxyFromCache(reference);
```
Comment (bentoi) [inline csharp/src/Ice/LocatorInfo.cs:107]: The "endpointsAge" variable name is not consistent with the naming of the return value field "LocationAge"

File: csharp/src/Ice/LocatorInfo.cs [hunk 2]
```diff
@@ -90,83 +90,81 @@ internal void ClearCache(Reference reference)
...
                 Debug.Assert(endpoints.Count == 0);
 
-                bool cachedLocation;
-
-                (endpoints, cachedLocation) = GetResolvedLocationFromCache(location,
-                                                                           reference.Protocol,
-                                                                           reference.LocatorCacheTimeout);
-
-                if (!cachedLocation)
+                (endpoints, locationAge) = GetResolvedLocationFromCache(location,
```
Comment (bentoi) [inline csharp/src/Ice/LocatorInfo.cs:132]: Shouldn't locationAge be endpointsAge?

File: csharp/src/Ice/LocatorInfo.cs [hunk 3]
```diff
@@ -90,83 +90,81 @@ internal void ClearCache(Reference reference)
...
+                                                                        reference.Protocol,
+                                                                        reference.LocatorCacheTimeout);
+                bool expired = CheckExpired(locationAge, reference.LocatorCacheTimeout);
+                if (endpoints.Count == 0 ||
+                    (expired && !_background) ||
+                    (endpointsMaxAge != Timeout.InfiniteTimeSpan && (locationAge > endpointsMaxAge || endpointsAge <= locationAge)))
```
Comment (bentoi) [inline csharp/src/Ice/LocatorInfo.cs:138]: Is `endpointsMaxAge != Timeout.InfiniteTimeSpan` really needed here? Comparing locationAge with endpointsAge is also questionable, is it really important?

Comment (bentoi) [inline csharp/src/Ice/LocatorInfo.cs:138]: Shouldn't it be `locationAge >= endpointsMaxAge` here as well or otherwise it won't retrieve new endpoints when called with the previous endpoint age?

File: csharp/src/Ice/LocatorInfo.cs [hunk 4]
```diff
@@ -90,83 +90,81 @@ internal void ClearCache(Reference reference)
...
-                                                                                   reference.LocatorCacheTimeout);
-                if (!cached)
+                (endpoints, location, endpointsAge) = GetResolvedWellKnownProxyFromCache(reference);
+                bool expired = CheckExpired(endpointsAge, reference.LocatorCacheTimeout);
+                // If no endpoints are returned from the cache, or if the cache returned and expired endpoint and
+                // background updates are disabled, or if the caller is requesting a more recent endpoint that the
+                // one returned from the cache, we try to resolve the endpoint again.
+                if ((endpoints.Count == 0 && location.Count == 0) ||
+                    (!_background && expired) ||
+                    (endpointsMaxAge != Timeout.InfiniteTimeSpan && endpointsAge >= endpointsMaxAge))
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/LocatorInfo.cs:114]: The first call pass `endpointsMaxAge = Timeout.InfiniteTimeSpan` and if we don't have this check it will query the locator even if the cached endpoint is still valid

[the remark this answers is not in the document]
Comment (bentoi) [reply csharp/src/Ice/LocatorInfo.cs:114]: Ok, `Timeout.InfiniteTimeSpan` is actually defined as `-1ms` so this is indeed necessary. Rather than calling this method with an infinite timeout, we could pass `TimeSpan.MaxValue` to indicate that we always want to use the cached value and we could get rid of this explicit check on `Timeout.InfiniteTimeSpan`.

[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/LocatorInfo.cs:114]: I was thinking it would be simpler to pass the Min value of LocatorCacheTimeout or the last resolved time from the caller, the difficulty is with background updates as we allow to use an expired endpoint and queue the update to run in the background.

[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/LocatorInfo.cs:114]: Note we also use -1 InfiniteTimeSpan for the cache timeout, that is the main reason I pick that, it seems a bit more consistent.

File: csharp/src/Ice/LocatorInfo.cs [hunk 5]
```diff
@@ -90,83 +90,81 @@ internal void ClearCache(Reference reference)
...
-
             if (reference.IsWellKnown)
             {
                 // First, we check the cache.
-                (endpoints, location, cached) = GetResolvedWellKnownProxyFromCache(reference,
-                                                                                   reference.LocatorCacheTimeout);
-                if (!cached)
+                (endpoints, location, endpointsAge) = GetResolvedWellKnownProxyFromCache(reference);
+                bool expired = CheckExpired(endpointsAge, reference.LocatorCacheTimeout);
+                // If no endpoints are returned from the cache, or if the cache returned and expired endpoint and
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/LocatorInfo.cs]: Fixed

File: csharp/src/Ice/LocatorInfo.cs [hunk 6]
```diff
@@ -90,83 +90,81 @@ internal void ClearCache(Reference reference)
...
-                                                                           reference.Protocol,
-                                                                           reference.LocatorCacheTimeout);
-
-                if (!cachedLocation)
+                (endpoints, locationAge) = GetResolvedLocationFromCache(location,
+                                                                        reference.Protocol,
+                                                                        reference.LocatorCacheTimeout);
+                bool expired = CheckExpired(locationAge, reference.LocatorCacheTimeout);
+                if (endpoints.Count == 0 ||
+                    (expired && !_background) ||
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/LocatorInfo.cs]: Fixed

File: csharp/src/Ice/Reference.cs [hunk 7]
```diff
@@ -1415,24 +1418,25 @@ internal async ValueTask<Connection> GetConnectionAsync(CancellationToken cancel
...
+                (endpoints, endpointsAge) = await ComputeEndpointsAsync(endpointsMaxAge, cancel).ConfigureAwait(false);
                 connection = Communicator.GetConnection(endpoints, PreferNonSecure, Label);
                 if (CacheConnection)
                 {
                     _connection = connection;
                 }
             }
 
-            if (connection == null)
+            while (connection == null && endpointsMaxAge == Timeout.InfiniteTimeSpan)
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/Reference.cs]: Reworked this, it was not correct

File: csharp/src/Ice/LocatorInfo.cs (hunk 1 above)
Comment (pepone) [reply csharp/src/Ice/LocatorInfo.cs]: Fixed

File: csharp/src/Ice/LocatorInfo.cs (hunk 2 above)
Comment (pepone) [reply csharp/src/Ice/LocatorInfo.cs]: Fixed

File: csharp/src/Ice/LocatorInfo.cs (hunk 3 above)
Comment (pepone) [reply csharp/src/Ice/LocatorInfo.cs]: Yes, you are right!

File: csharp/src/Ice/LocatorInfo.cs [hunk 8]
```diff
@@ -90,83 +90,82 @@ internal void ClearCache(Reference reference)
...
+                    // Endpoints are returned from the cache but TTL was reached, if backgrounds updates
+                    // are configured, we obtain new endpoints but continue using the stale endpoints to
+                    // not block the caller.
+                    _ = ResolveLocationAsync(location, reference.Protocol, reference.Communicator, cancel: default);
+                }
 
                 if (!reference.IsWellKnown)
                 {
-                    cached = cachedLocation;
+                    locationAge = endpointsAge;
```
Comment (bentoi) [inline csharp/src/Ice/LocatorInfo.cs:168]: If this `if` still needed? Can't we just always return `endpointsAge` instead of `locationAge`?

**Review by bentoi — APPROVED** [review]
Looks good.

**Review by bernardnormier — COMMENTED** [review]
I find the logic hard to follow, with not enough comments.

Would it be possible to merge locatorCacheTimeout == TTL and maxAge into a single parameter/check?

File: csharp/src/Ice/LocatorInfo.cs [hunk 9]
```diff
@@ -313,7 +309,7 @@ private void ClearCache(Location location, Protocol protocol, Communicator commu
             }
         }
 
-        private (EndpointList Endpoints, bool Cached) GetResolvedLocationFromCache(
+        private (EndpointList Endpoints, TimeSpan EndpointsAge) GetResolvedLocationFromCache(
             Location location,
             Protocol protocol,
             TimeSpan ttl)
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/LocatorInfo.cs:315]: Only to keep returning empty enpoints when ttl/LocatorCacheTimeout is 0, maybe would be clear to not call this method when LocatorCacheTimeout is 0

**Comment by pepone** [comment]
> Would it be possible to merge locatorCacheTimeout == TTL and maxAge into a single parameter/check?

The difficulty for using a single check is the background updates, when we set maxAge we are requesting a fresh endpoint, and we wait for it to be resolved if required independently of the background update settings.

For the LocatorCacheTimeout if the endpoint has reached the maxAge/TTL, we still use it when the background updates are configured

File: csharp/src/Ice/Reference.cs [hunk 10]
```diff
@@ -1453,10 +1458,18 @@ internal async ValueTask<Connection> GetConnectionAsync(CancellationToken cancel
                     catch
                     {
                         // Ignore the exception unless this is the last endpoint.
-                        // TODO retry with non cached endpoints
                         if (ReferenceEquals(endpoint, last))
                         {
-                            throw;
+                            if (IsIndirect && endpointsMaxAge == TimeSpan.MaxValue)
+                            {
+                                endpointsMaxAge = endpointsAge;
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/Reference.cs:1465]: Yes, this avoids clearing the cache, as we were doing before.

[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/Reference.cs:1465]: Would be nice to add a comment.

**Review by bernardnormier — APPROVED** [review]
Looks good.
